### PR TITLE
[wip] enable sudo-less install for admins

### PR DESCRIPTION
**Pull request by minrk**
Goal: members of jupyterhub-admins should be able to install and upgrade packages without sudo.

The main change is using `setfacl` and `setgid` on the top-level directory to ensure that new files and directories in the env are owned by `jupyterhub-admins` and group-writable.

Changes:

- merge hub and user env in `/opt/tljh/env`
- move state out of env to `/opt/tljh/state`
- integration tests for permissions:
  - everyone can read every file in the env
  - admins can write any directory
  - neither admins nor users can read or write to state
  - admins can pip install/upgrade
  - users cannot pip install/upgrade

Somewhat unrelated fixes along the way:

- allow conda to be upgraded (fix == version check instead of >= check)

Some weirdnesses:

Not all *files* are group-writable. This is because of the behavior of pip and conda installers, which can make files read-only and/or not group-owned. This is okay for our purposes, because installation/upgrade should never involve *modification* of files, only removal of files and writing of new files. That means that only write-permission on directories is required. Sudo remains available if individual files want to be modified, which should be a rare, advanced use case.

### CODE REVIEW

**Comment by yuvipanda** [comment]
Thanks for working on this, @minrk! <3

I'd like to keep the user and hub env separate, primarily as a protection against users accidentally screwing up the hub environment. Especially since the primary way we want users to interact with the hub is via the hub itself, this could be bad.

I'll take a more detailed look tomorrow!

File: .circleci/config.yml [hunk 1]
```diff
@@ -37,49 +37,61 @@ jobs:
                     codecov
 
     integration-test:
-        docker:
-            - image: docker:18.05.0-ce-git
-
+        machine: true
```
Comment (yuvipanda) [inline .circleci/config.yml]: is there a reason this is using machine instead of container?

Comment (yuvipanda) [reply .circleci/config.yml]: ah, I see you're just hacking around it to find out why things are failing :)

**Comment by yuvipanda** [comment]
I'm also now less sure that the amount of complexity (around ACLs and sticky bits) this adds is worth the effort at this stage of the project. How do you feel about delaying this until we get a release out later the coming week?

File: .circleci/config.yml (hunk 1 above)
Comment (minrk) [reply .circleci/config.yml]: Yup. While the acls work on a 'real' machine, the filesystem on CI has some of the same acl issues that images have. After spending a bit of time, I think we can't reasonably test on Circle if we want to see how a real filesystem would be have.

**Comment by minrk** [comment]
> I'd like to keep the user and hub env separate

Sure! I thought we discussed this yesterday and concluded that the envs should not be separate, but that's easy to revert.

After fighting with Circle for a bit and dealing with weird issues, I think I'll break this up into one or two smaller PRs that are mostly the fixes-along-the-way instead of the original motivating feature.

> How do you feel about delaying this until we get a release out later the coming week?

Makes perfect sense. I think one or two of the smaller fixes should get in, but I think it's a good idea to save the "admin install without sudo" feature for later.

I really wish people cared about filesystems in containers :(.

**Comment by minrk** [comment]
No need to keep this open, since it's pretty hairy to test (easier to do in practice, but CircleCI doesn't give us a realistic environment in which to test).

**Pull request closed without merging**